collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java

```diff
@@ -198,7 +231,99 @@ private String getCurrentPath() {
+            return path;
+        }
+    }
+
+    /**
+     * Goes to the start of the hierarchy view based on where the user came from.
+     * Backs out until the index is at the beginning of a repeat group or the beginning of the form.
+     */
+    private void jumpToHierarchyStartIndex() {
+        FormController formController = Collect.getInstance().getFormController();
+        FormIndex startIndex = formController.getFormIndex();
+
+        // If we're not at the first level, we're inside a repeated group so we want to only
+        // display everything enclosed within that group.
+        contextGroupRef = "";
+
+        // Save the index to the screen itself, before potentially moving into it.
+        screenIndex = startIndex;
+
+        // If we're currently at a repeat node, record the name of the node and step to the next
+        // node to display.
+        if (formController.getEvent() == FormEntryController.EVENT_REPEAT) {
+            contextGroupRef = getGroupRef(formController);
+            formController.stepToNextEvent(FormController.STEP_INTO_GROUP);
+        } else {
+            FormIndex potentialStartIndex = formController.stepIndexOut(startIndex);
+            // Step back until we hit a repeat or the beginning.
+            while (!isScreenEvent(formController, potentialStartIndex)) {
+                potentialStartIndex = formController.stepIndexOut(potentialStartIndex);
+            }
+
+            screenIndex = potentialStartIndex;
+
+            if (potentialStartIndex == null) {
+                // check to see if the question is at the first level of the hierarchy. If it
+                // is, display the root level from the beginning.
+                formController.jumpToIndex(FormIndex.createBeginningOfFormIndex());
+            } else {
+                // otherwise we're at a repeated group
+                formController.jumpToIndex(potentialStartIndex);
+            }
+
+            int event = formController.getEvent();
+
+            // now test again for repeat. This should be true at this point or we're at the
+            // beginning
+            if (event == FormEntryController.EVENT_REPEAT) {
+                contextGroupRef = getGroupRef(formController);
+                formController.stepToNextEvent(FormController.STEP_INTO_GROUP);
+            }
+        }
+    }
+
+    /**
+     * Returns true if the event is an enclosing repeat or the start of the form.
+     * See {@link FormController#stepToOuterScreenEvent} for more context.
+     */
+    private boolean isScreenEvent(FormController formController, FormIndex index) {
+        // Beginning of form.
+        if (index == null) {
+            return true;
+        }
+
+        int event = formController.getEvent(index);
+        return event == FormEntryController.EVENT_REPEAT;
+    }
+
+    private String getGroupRef(FormController formController) {
+        return formController.getFormIndex().getReference().toString(true);
```

Comment:
You can call just toString() because it uses true by default then:
https://github.com/opendatakit/javarosa/blob/master/src/org/javarosa/core/model/instance/TreeReference.java#L472